File: tests/_evil_worker.py
"""Adversarial channel peer used by transport tests (Voss RFC 9.2).

Run as ``python -m tests._evil_worker`` with ``VOSS_CHANNEL_BOOTSTRAP`` set
(the runtime provisions it exactly like a real worker).  ``VOSS_EVIL_MODE``
selects the attack:

- handshake-forged-mac       bad HMAC on the hello record
- reply-forged-mac           bad HMAC on the proposal record
- reply-replay               resend the (valid) hello record: duplicate seq
- reply-skip-seq             jump the sequence forward
- reply-wrong-direction      send a host-bound record on the adapter link
- reply-oversize             a line longer than the transport cap

The trusted runtime must refuse every one of these with a precise reason and
contain the worker.
"""
from __future__ import annotations

import json
import os
import sys

import voss.chan as chan

MAX = chan.MAX_LINE


def _corrupt_mac(wire_line: str) -> str:
    record = json.loads(wire_line)
    record["mac"] = "0" * 64
    return json.dumps(record, sort_keys=True)


def main() -> int:
    mode = os.environ.get("VOSS_EVIL_MODE", "valid")
    try:
        key, sid = chan.read_bootstrap_file(os.environ["VOSS_CHANNEL_BOOTSTRAP"])
    except Exception as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 2

    session = chan.ChannelSession(key, sid, "adapter")

    if mode == "handshake-forged-mac":
        print(_corrupt_mac(
            session.send("hello", {"adapter": "evil", "ready": True})),
            flush=True)
        return 0

    valid_hello = session.send("hello", {"adapter": "evil", "ready": True})
    print(valid_hello, flush=True)

    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue
        try:
            msg_type, msg = session.receive(line)
        except chan.ChannelError:
            return 3
        if msg_type != "prompt":
            continue

        if mode == "reply-forged-mac":
            print(_corrupt_mac(session.send("proposal", {"envelopes": []})),
                  flush=True)
        elif mode == "reply-replay":
            print(valid_hello, flush=True)  # duplicate a2h seq=1
        elif mode == "reply-skip-seq":
            print(chan.wire_line(
                key, sid, "a2h", 3, "proposal", {"envelopes": []}), flush=True)
        elif mode == "reply-wrong-direction":
            print(chan.wire_line(
                key, sid, "h2a", 1, "prompt",
                {"prompt": "", "session_id": "", "principal": ""}),
                flush=True)
        elif mode == "reply-oversize":
            print("x" * (MAX + 1), flush=True)
        else:  # valid
            print(session.send("proposal", {"envelopes": []}), flush=True)
        return 0
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: rust/src/bin/evil_worker.rs
//! Adversarial channel peer for transport tests.
//!
//! `VOSS_EVIL_MODE` selects the attack. The runtime provisions the bootstrap
//! file the same way it does for a real worker.

use std::io::{self, BufRead, Write};
use std::path::Path;

use voss::canonical::{Json, canonical_bytes, loads_strict};
use voss::chan::{self, ChannelSession, MAX_LINE};

fn main() {
    let code = run();
    std::process::exit(code);
}

fn run() -> i32 {
    let mode = std::env::var("VOSS_EVIL_MODE").unwrap_or_else(|_| "valid".to_string());
    let path = match std::env::var("VOSS_CHANNEL_BOOTSTRAP") {
        Ok(path) if !path.is_empty() => path,
        _ => {
            eprintln!("error: denied_channel_bootstrap");
            return 2;
        }
    };
    let (key, sid) = match chan::consume_bootstrap(Path::new(&path)) {
        Ok(value) => value,
        Err(error) => {
            eprintln!("error: {}", error.code());
            return 2;
        }
    };
    let mut session = match ChannelSession::new(key.clone(), sid.clone(), "adapter") {
        Ok(session) => session,
        Err(error) => {
            eprintln!("error: {}", error.code());
            return 2;
        }
    };
    if mode == "handshake-forged-mac" {
        let hello = match session.send("hello", &hello_body()) {
            Ok(line) => line,
            Err(error) => {
                eprintln!("error: {}", error.code());
                return 3;
            }
        };
        println!("{}", corrupt_mac(&hello));
        let _ = io::stdout().flush();
        return 0;
    }
    let valid_hello = match session.send("hello", &hello_body()) {
        Ok(line) => line,
        Err(error) => {
            eprintln!("error: {}", error.code());
            return 3;
        }
    };
    println!("{valid_hello}");
    let _ = io::stdout().flush();

    let stdin = io::stdin();
    for line in stdin.lock().lines() {
        let Ok(line) = line else {
            break;
        };
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        match session.receive(line) {
            Err(_) => return 3,
            Ok((msg_type, _)) if msg_type != "prompt" => continue,
            Ok(_) => {}
        }
        let reply = match mode.as_str() {
            "reply-forged-mac" => {
                let proposal = match session.send("proposal", &empty_proposal()) {
                    Ok(line) => line,
                    Err(_) => return 3,
                };
                corrupt_mac(&proposal)
            }
            "reply-replay" => valid_hello.clone(),
            "reply-skip-seq" => match chan::wire_line(&key, &sid, "a2h", 3, "proposal", &empty_proposal()) {
                Ok(line) => line,
                Err(_) => return 3,
            },
            "reply-wrong-direction" => {
                let msg = Json::object([
                    ("prompt", Json::string("")),
                    ("session_id", Json::string("")),
                    ("principal", Json::string("")),
                ]);
                match chan::wire_line(&key, &sid, "h2a", 1, "prompt", &msg) {
                    Ok(line) => line,
                    Err(_) => return 3,
                }
            }
            "reply-oversize" => "x".repeat(MAX_LINE + 1),
            _ => match session.send("proposal", &empty_proposal()) {
                Ok(line) => line,
                Err(_) => return 3,
            },
        };
        println!("{reply}");
        let _ = io::stdout().flush();
        return 0;
    }
    0
}

fn hello_body() -> Json {
    Json::object([
        ("adapter", Json::string("evil")),
        ("ready", Json::Bool(true)),
    ])
}

fn empty_proposal() -> Json {
    Json::object([("envelopes", Json::Array(Vec::new()))])
}

fn corrupt_mac(wire_line: &str) -> String {
    let Ok(mut record) = loads_strict(wire_line) else {
        return wire_line.to_string();
    };
    let Json::Object(fields) = &mut record else {
        return wire_line.to_string();
    };
    fields.insert("mac".to_string(), Json::string("0".repeat(64)));
    canonical_bytes(&record)
        .ok()
        .and_then(|bytes| String::from_utf8(bytes).ok())
        .unwrap_or_else(|| wire_line.to_string())
}
